### PR TITLE
[Feat] Add @OptionFunction decorator

### LINKED ISSUE #28: [RFC] Function decoding in configuration
Since we have #26 completed all configuration is now decoded via `io-ts` codecs which allows us to implement custom types.

## Description

Create a type that can decode stringified (serialised) function into valid JS function for any configuration property.

## Behavior

- Custom `io-ts` type `FunctionFromString` that will accept 1 of the next inputs:
  1. serialised function
  2. object `{args: string[], body: string}`, where `args` is a list of function's arguments and `body` is function's body

- If input is invalid - validation will fail

- If input is accepted then type will, in case:
  1. Try to parse input string to extract function's arguments and body and then construct new `Function` based on that info.
  In case of failure - validation will fail.
  2. construct new `Function` based on input object.

- When function is constructed it's arguments should be resolved by the map `INJECTOR_MAP` via local `Injector` and bound to function.
- **NEW** Arguments whose name starts with `$` must be:
  - Skipped by injector resolution algorithm
  - Moved to the end of the arguments list saving their relative order to same arguments (`arg1, $arg2, arg3, $arg4` => `arg1, arg3, $arg2, $arg4`) so that they will not be bound and accepted from caller.
- `INJECTOR_MAP` is a `multi provider` of type `{[key: string]: Type<any> | InjectionToken<any>}`.
  In this way any external package can provide custom extra mappings for their functions to inject.
- If some arguments were not resolved by `Injector` we will use default value if provided, otherwise the error will be thrown.
- Errors from function will be captured, wrapped in `FunctionError` and handled via `ErrorStrategy`.

_Extra:_
- `INJECTOR_MAP` by default will contain mappings for core services and components:
  - `getInjector()` - will return local `Injector` with mapping capability built-in (same instance used to resolve function's arguments)
  - `InjectFlags` - will return `InjectFlags` enum from `@angular/core` (useful  to use with `getInjector().get(...)`)
  - `getComponent()` - will return closest dynamic component instance
  - `getConfiguration()` - will return closest dynamic component configuration object
  - `isConfigurationValid()` - will return boolean telling the validation status of configuration of closest dynamic component
  - `getRenderItem()` - will return closest `RenderItemComponent` instance

### `@OptionFunction`
```ts
@OptionFunction(customInjector?: CustomInjectorFactory)
```
> - Will enable validation as well as decoding of serialised function.
> - Optional `CustomInjectorFactory` can be supplied to provide custom arguments resolution of specific function's arguments.

### `CustomInjectorFactory`
```ts
import { Injector } from '@angular/core';
type CustomInjectorFactory = (parentInjector: Injector) => Injector;
```
> - Accepts parent injector and returns configured new `Injector`

## Example

```ts
// Define config with function property
class MyConfig {
  @OptionFunction()
  fn: Function;
}

// Decode and validate config at runtime
const myConfig = this.configurationService.decode(
  MyConfig,
  { fn: '() => "hi from custom function!"' },
  customInjector // This must be `local Injector` that will be used to resolve any function's arguments
);

// Use decoded config
myConfig.fn(); // Will return `hi from custom function!`
```

/cc @RyuuGan @baruchvlz @utkuturunc what do you think? anything has to be adjusted/added?

### PR DESCRIPTION
**Pull request by gund**
Implementation of RFC #28 (closes #28)

- [x] Decode/validate serialized function
- [x] Decode/validate function metadata (`{args: string[], body: string}`)
- [x] Resolve arguments via local `Injector`
- [x] Capture errors from functions and forward to `ErrorStrategy`
- [x] Handle arguments which starts with `$`

### CODE REVIEW

File: libs/core/src/lib/util.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  while ((group = regex.exec(val)) !== null) {
+    if (group.index === regex.lastIndex) {
+      regex.lastIndex++;
+    }
+
+    arr.push(...group);
+  }
+
+  return !arr.length ? null : arr;
```
Comment (PxyUp) [inline libs/core/src/lib/util.ts]: May be here return always arr, instead return null in case arr.length === 0

Comment (gund) [reply libs/core/src/lib/util.ts]: This is done to copy same behavior of native `match/exec` methods, if nothing was found - `null` will be returned.

**Comment by gund** [comment]
Implementation is complete. More unit tests required to verify that it's working.

**Comment by gund** [comment]
Implementation is complete, need more unit tests.

**Comment by gund** [comment]
Implementation and unit tests are complete now.

/cc @RyuuGan @baruchvlz please review before merging.

**Comment by baruchvlz** [comment]
Shouldn't we move the UI stuff to another PR? I get that it doesn't really matter much, but you know, keep things separated.

**Comment by gund** [comment]
@baruchvlz yeah you are right. We should, but here I just wanted to quickly verify in demo app that it is working end-to-end so I had to add just that little bit.

I won't do it again =)

**Comment by gund** [comment]
I've added a new prop to dynamic component's that allows to set any [event handlers on host components](https://github.com/orchestratora/orchestrator/pull/29/files#diff-e781977c32f0e7ab20244dfc0e965aabR18).

Thanks to this I removed `click` from ui-web-button component config.

**Pull request merged**